src/dsync/dsync.c

```diff
@@ -1305,9 +1325,17 @@ static int dsync_sync_files(
         if(rank == 0) {
             MFU_LOG(MFU_LOG_ERR, "Exiting run");
         }
-        mfu_finalize();
-        MPI_Finalize();
-        return 0;
+         /* TODO this is problematic for a few reasons:
+         * 1. The calling function doesn't see this as an error.
+         * 2. The calling function still tries to use MPI.
+         * 3. Resources are not cleaned up. */
+        //mfu_finalize();
```

Comment:
@daltonbohning did you mean to leave in the commented out functions? So, it looks like before that if MPI_Finalize is called are you saying that it still tries to use MPI afterward?

Reply:
@dsikich Yeah, really I was hoping Adam might have some input here. The problem is that this is not in main, so finalizing shouldn't be done here. "return 0" just returns from the function, which still expects the MPI context to be initialized.

Reply:
@daltonbohning @adammoody ok, yeah this is kind of weird.

Reply:
I'm hoping that maybe the "return -1" I added is sufficient, though really it seems that this is a "fatal" error, so the caller should also immediately return, and the caller's caller also... which is what makes this complex.
So really, if we can turn this into a "log error and continue" instead of "exit immediately", the "return -1" should be okay.

Reply:
@daltonbohning yes I think it should be ok, but we will probably want to delete the commented out functions if we decide to do it that way.

Reply:
That "exit when invalid" code probably came in from a cut-and-paste when moving this logic from the main function of the dsync tool into a function.  Yes, we should just have the function return an error in this case.  Thanks @daltonbohning and @dsikich

Reply:
The goal was to eventually move this dsync_sync_files function into libmfu as mfu_flist_sync that other tools could call.  I think it needed more work, and that I stopped half way after pulling code from the dsync main() function into separate functions.

Looking at dsync_sync_files a bit closer, let's ensure that it returns the same value on all ranks (global return value).  I think we can just have this part return -1, assuming all ranks get ``!valid`` at the same time.  I don't remember off hand about other points where we set ``rc = -1`` further down in that function.  To be safe, we could add a ``mfu_alltrue`` call before returning.

Then in the caller, my first reaction was that we should avoid setting metadata on failure.  However, now I'm not sure.  I could see that we might get some errors when sync'ing files, because perhaps someone deleted a file while the tool was running.  In that case, they probably would want to go ahead and set metadata on things.

Reply:
@adammoody `dsync_sync_files` sets `rc = -1` for these cases:
- `mfu_flist_hardlink` - safe - Performs MPI_Allreduce before returning
- `mfu_flist_copy` - safe - Performs MPI_Allreduce before returning

And for `mfu_param_path_check_copy` (the one in question here), rank 0 performs a bcast into `valid`.
So I think the return from `dsync_sync_files` should be the same for all processes.

For now, I kept the `return -1` for `mfu_param_path_check_copy`, and kept a TODO comment in place so we can keep track of the potential improvement here.

As for setting metadata, I agree. There are some cases where maybe we should and some cases where maybe we should not. For this particular case - `mfu_param_path_check_copy` error, it's probably okay to NOT set the metadata, since it's possible that the source or destination path itself is invalid. But I think this would require more careful analysis, and possibly some refactoring as you suggested.